File: src/main/java/com/google/android/finsky/api/DfeRetryPolicy.java
package com.google.android.finsky.api;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.VolleyError;

public class DfeRetryPolicy extends DefaultRetryPolicy {
    private final DfeApiContext mDfeApiContext;
    private boolean mHadAuthException;


    public DfeRetryPolicy(int initialTimeoutMs, int maxNumRetries, float backoffMultiplier, final DfeApiContext mDfeApiContext) {
        super(initialTimeoutMs, maxNumRetries, backoffMultiplier);
        this.mDfeApiContext = mDfeApiContext;
    }

    public DfeRetryPolicy(final DfeApiContext mDfeApiContext) {
        super();
        this.mDfeApiContext = mDfeApiContext;
    }

    @Override
    public void retry(final VolleyError volleyError) throws VolleyError {
        if (volleyError instanceof AuthFailureError) {
            if (this.mHadAuthException) {
                throw volleyError;
            }
            this.mHadAuthException = true;
        }
        super.retry(volleyError);
    }
}

File: playstore/src/main/java/finsky/api/DfeRetryPolicy.kt
package finsky.api

import com.android.volley.AuthFailureError
import com.android.volley.DefaultRetryPolicy
import com.android.volley.VolleyError

class DfeRetryPolicy : DefaultRetryPolicy {
    private val dfeApiContext: DfeApiContext
    private var hadAuthException: Boolean = false

    constructor(initialTimeoutMs: Int, maxNumRetries: Int, backoffMultiplier: Float, dfeApiContext1: DfeApiContext)
            : super(initialTimeoutMs, maxNumRetries, backoffMultiplier) {
        this.dfeApiContext = dfeApiContext1
    }

    constructor(dfeApiContext: DfeApiContext) : super() {
        this.dfeApiContext = dfeApiContext
    }

    @Throws(VolleyError::class)
    override fun retry(volleyError: VolleyError?) {
        if (volleyError is AuthFailureError) {
            if (this.hadAuthException) {
                throw volleyError
            }
            this.hadAuthException = true
        }
        super.retry(volleyError)
    }
}